src/annotator/anchoring/html.js

```diff
@@ -0,0 +1,124 @@
+import { RangeAnchor, TextPositionAnchor, TextQuoteAnchor } from './types';
+
+/**
+ * @typedef {import("./types").AnyRangeType} AnyRangeType
+ * @typedef {import('../../types/api').Selector} Selector
+ */
+
+/**
+ * @param {RangeAnchor|TextPositionAnchor|TextQuoteAnchor} anchor
+ * @param {Object} [options]
+ *  @param {number} [options.hint]
+ */
+const querySelector = function (anchor, options = {}) {
+  const doQuery = function (resolve, reject) {
+    try {
+      const range = anchor.toRange(options);
+      return resolve(range);
+    } catch (error) {
+      return reject(error);
+    }
+  };
+  return new Promise(doQuery);
+};
+
+/**
+ * Anchor a set of selectors.
+ *
+ * This function converts a set of selectors into a document range.
+ * It encapsulates the core anchoring algorithm, using the selectors alone or
+ * in combination to establish the best anchor within the document.
+ *
+ * @param {Node} root - The root element of the anchoring context.
+ * @param {Selector[]} selectors - The selectors to try.
+ * @param {Object} [options]
+ *   @param {number} [options.hint]
+ */
+export function anchor(root, selectors, options = {}) {
+  let position = null;
+  let quote = null;
+  let range = null;
+
+  // Collect all the selectors
+  for (let selector of selectors) {
+    switch (selector.type) {
+      case 'TextPositionSelector':
+        position = selector;
+        options.hint = position.start; // TextQuoteAnchor hint
+        break;
+      case 'TextQuoteSelector':
+        quote = selector;
+        break;
+      case 'RangeSelector':
+        range = selector;
+        break;
+    }
+  }
+
+  /**
+   * Assert the quote matches the stored quote, if applicable
+   * @param {Range} range
+   */
+  const maybeAssertQuote = range => {
+    if (quote?.exact && range.toString() !== quote.exact) {
+      throw new Error('quote mismatch');
+    } else {
+      return range;
+    }
+  };
+
+  // From a default of failure, we build up catch clauses to try selectors in
+  // order, from simple to complex.
+  let promise = Promise.reject('unable to anchor');
```

Comment:
If you widen the type of `promise` then you can remove the `// @ts-ignore` comments below:

```js
  /** @type {Promise<Range>} */
  let promise = Promise.reject('unable to anchor');
```